File: test/e2e/Apps/BasicNode/src/functions/SuspendResumeOrchestration.ts
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { app, HttpHandler, HttpRequest, HttpResponseInit, InvocationContext } from '@azure/functions';
import * as df from 'durable-functions';
import { DurableClient } from 'durable-functions';

// SuspendInstance HTTP trigger
const SuspendInstance: HttpHandler = async (request: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> => {
    const client: DurableClient = df.getClient(context);
    const instanceId = request.params.instanceId;
    const suspendReason = "Suspending the instance for test.";
    try {
        // Reason for the cast - Bug: https://github.com/Azure/azure-functions-durable-js/issues/608
        await (client as any).suspend(instanceId, suspendReason);
        return { status: 200 };
    } catch (ex: any) {
        context.log(`Error suspending instance: ${ex}`);
        return {
            status: 400,
            body: String(ex),
            headers: { "Content-Type": "text/plain" }
        };
    }
};

app.http('SuspendInstance', {
    route: 'SuspendInstance',
    methods: ['GET', 'POST'],
    extraInputs: [df.input.durableClient()],
    handler: SuspendInstance,
});

// ResumeInstance HTTP trigger
const ResumeInstance: HttpHandler = async (request: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> => {
    const client = df.getClient(context);
    const instanceId = request.params.instanceId;
    const resumeReason = "Resuming the instance for test.";
    try {
        // Reason for the cast - Bug: https://github.com/Azure/azure-functions-durable-js/issues/608
        await (client as any).resume(instanceId, resumeReason);
        return { status: 200 };
    } catch (ex: any) {
        context.log(`Error resuming instance: ${ex}`);
        return {
            status: 400,
            body: String(ex),
            headers: { "Content-Type": "text/plain" }
        };
    }
};

app.http('ResumeInstance', {
    route: 'ResumeInstance',
    methods: ['GET', 'POST'],
    extraInputs: [df.input.durableClient()],
    handler: ResumeInstance,
});

File: test/e2e/Apps/BasicJava/src/main/java/com/function/SuspendResumeOrchestration.java
package com.function;

import com.microsoft.azure.functions.annotation.*;
import com.microsoft.azure.functions.*;

import com.microsoft.durabletask.*;
import com.microsoft.durabletask.azurefunctions.DurableClientContext;
import com.microsoft.durabletask.azurefunctions.DurableClientInput;

import java.util.Optional;


/**
 * Azure Functions with HTTP Trigger for suspending and resuming orchestration instances.
 */
public class SuspendResumeOrchestration {

    /**
     * HTTP-triggered function to suspend an orchestration instance.
     */
    @FunctionName("SuspendInstance")
    public HttpResponseMessage suspendInstance(
            @HttpTrigger(name = "req", methods = {HttpMethod.GET, HttpMethod.POST}, authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<Optional<String>> request,
            @DurableClientInput(name = "durableContext") DurableClientContext durableContext,
            @BindingName("instanceId") String instanceId,
            final ExecutionContext context) {

        String suspendReason = "Suspending the instance for test.";
        DurableTaskClient client = durableContext.getClient();

        try {
            client.suspendInstance(instanceId, suspendReason);
            return request.createResponseBuilder(HttpStatus.OK).build();
        } catch (Exception ex) {
            HttpResponseMessage.Builder responseBuilder = request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                .header("Content-Type", "text/plain")
                .body(ex.getMessage());
            return responseBuilder.build();
        }
    }

    /**
     * HTTP-triggered function to resume an orchestration instance.
     */
    @FunctionName("ResumeInstance")
    public HttpResponseMessage resumeInstance(
            @HttpTrigger(name = "req", methods = {HttpMethod.GET, HttpMethod.POST}, authLevel = AuthorizationLevel.ANONYMOUS) HttpRequestMessage<Optional<String>> request,
            @DurableClientInput(name = "durableContext") DurableClientContext durableContext,
            @BindingName("instanceId") String instanceId,
            final ExecutionContext context) {

        String resumeReason = "Resuming the instance for test.";
        DurableTaskClient client = durableContext.getClient();

        try {
            client.resumeInstance(instanceId, resumeReason);
            return request.createResponseBuilder(HttpStatus.OK).build();
        } catch (Exception ex) {
            HttpResponseMessage.Builder responseBuilder = request.createResponseBuilder(HttpStatus.BAD_REQUEST)
                .header("Content-Type", "text/plain")
                .body(ex.getMessage());
            return responseBuilder.build();
        }
    }
}